File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        
        ListNode* curr1 = l1;
        ListNode* curr2 = l2;
        
        ListNode* head = new ListNode(0);
        ListNode* currh = head;
        
        int carry = 0;
        
        
        while (curr1 != NULL || curr2 != NULL || carry != 0) {
            int sum = carry;
            
            if (curr1 != NULL) {
                sum += curr1->val;
                curr1 = curr1->next;
            }
            if (curr2 != NULL) {
                sum += curr2->val;
                curr2 = curr2->next;
            }
            
            carry = sum/10;
            sum %= 10;
            
            currh->next = new ListNode(sum);
            currh = currh->next;
        }
        
        return head->next;
        
    }
};

File: 0002-add-two-numbers/0002-add-two-numbers.scala
/**
 * Definition for singly-linked list.
 * class ListNode(_x: Int = 0, _next: ListNode = null) {
 *   var next: ListNode = _next
 *   var x: Int = _x
 * }
 */
object Solution {
    def addTwoNumbers(l1: ListNode, l2: ListNode): ListNode = {
        
        def solve(n1: ListNode, n2: ListNode, carry: Int): ListNode = {
            // Base Case: Both lists are empty and no carry remains
            if (n1 == null && n2 == null && carry == 0) {
                null
            } else {
                // Extract values (handle nulls if lists are different lengths)
                val v1 = if (n1 != null) n1.x else 0
                val v2 = if (n2 != null) n2.x else 0
                
                val sum = v1 + v2 + carry
                val newNode = new ListNode(sum % 10)
                
                // Recursive step: Move to the next nodes and pass the new carry
                newNode.next = solve(
                    if (n1 != null) n1.next else null,
                    if (n2 != null) n2.next else null,
                    sum / 10
                )
                
                newNode
            }
        }
        
        solve(l1, l2, 0)
        
    }
}